File: logistic_regression.py
#Logistic Regression 

#You want to highlight and run the code bits at a time
#import the libraries we need
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#make sure in the correct directory
#import the dataset

dataset = pd.read_csv('Social_Network_Ads.csv')

X = dataset.iloc[:, [2,3]].values #Matrix of current independent variables
y = dataset.iloc[:, -1].values # dependent variable...also could have used y = dataset.iloc[:, 4].values

from sklearn.cross_validation import train_test_split #i call the line of code below, the wendys 4for4
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0) 
#Are we going to apply feature scaling? Yes because we want to predict which users are going to buy the 
# SUV so that we can target these users as well as possible! Data Science!

# Remember, what we are doing is making sure that nothing off scale...meaning; 
# we dont want independent variables outweighing
# other independent variables in a way that they are not supposed to.
# To do that:
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train) #let the training set be equal to sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test) # let the test set be equal to sc_X.fit_transform(X_test)

#this two lines above really looks like StandardScaler().sc_X.fit_transform(X_train)
#Notice we go from class --> object of that class --> instance of that class


#-----Now we are going to fit the logistic regression models-----
from sklearn.linear_model import LogisticRegression #importing our LogisticRegression class
classifier = LogisticRegression(random_state = 0) #creating our -->>CLASSIFIER OBJECT<<--- from the class
# LogisticRegression using sklearn.linear_model
classifier.fit(X_train, y_train)   # fitting the classifier to our TRAINING set...
#so we are taking our LogisticRegression classifier object and we are FITTING it to the TRAINING set, X_train, y _train

#Now we are going to predict the test set results
y_pred = classifier.predict(X_test) # so we are going to take our FITTED classifier object from the line of code above 

#We are going to make something new here called the CONFUSION MATRIX
#We are going to import a function from the sklearn.metrics library to compute the CONFUSION MATRIX faster!
from sklearn.metrics import confusion_matrix 
#notice the lower cases here because we are importing a function not a class. A class contains CAPITALS

cm = confusion_matrix(y_test, y_pred) #we used to the confusion matrix to evaluate the 
# predictive power of our logistic regression model

#first parameter is y_true, that is the REAL values that happended in reality
#second parameter is y_pred, that is the vector of predictions that our logistic regression model predicted
#other two parameters don't need for now


# Now we are going to have a graphic visualization in our results
# This graph is going to be pretty good.
%matplotlib qt
from matplotlib.colors import ListedColormap #helps colorize all of the data points
X_set, y_set = X_train, y_train #create some local values so we can create a shortcut so that we dont have to go around and 
#replace every single X_test and y_test. ;) 

#Line X1,X2 we take the minimum ages and estimatedsalaries - 1 and + 1 because we dont want the data points to be squeezed in
#the graph and make it look ugly...we want some border room pretty much
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01), 
                    

                     np.arange(start = X_set[:,1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01)) 
                            #  #We took all of the pixel points available, and then we applied our classifier
                            # on it...so we just made a bunch of fake predictions
                            # and then applied our classifier on it...so its like each of the pixel points in     
                            # the graph is a user of the social network (like the solid red and green)
                            #Its like an observation, only its not one of the observations we had in our dataset


plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),  #this line here is where we 
             #actually apply the classifier to all of the "pixel observation points" and by doing that, 
             #it colorizes everything.. use contour function to actually make the contour between the two prediction regions.
            
             alpha = 0.75, cmap = ListedColormap(('red','green'))) #------>if the pixel point belongs to class zero  
                                                                    # its gets red, else it gets green

plt.xlim(X1.min(), X1.max()) #here we are plotting the limits of                               

plt.ylim(X2.min(), X2.max() #x, the age, and y, the estimatedsalary                              

for i, j in enumerate(np.unique(y_set)): #then with this for loop, we are plotting all of the data points that are the real 
                                        #values. (unique, separated)

    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1], #using a scatter plot for the REAL value

                c = ListedColormap(('red','green'))(i), label = j) #green they bought the product,  red they did not

#Just labeling the axes, providing a legend, and showing the graph plt.show()
plt.title('Logistic Regression ( Training Set )')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()


#Visualizing the Test set results below
%matplotlib qt
from matplotlib.colors import ListedColormap
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                    np.arange(start = X_set[:,1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
            alpha = 0.75, cmap = ListedColormap(('red','green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('red','green'))(i), label = j)

#Just labeling the axes, providing a legend, and showing the graph plt.show()
plt.title('Logistic Regression ( Test Set )')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()
""" 
We have some red points and green points.

All of the points you see on the graph are the observations from the TRAINING set.
That is, these are all of the users of the social network that were selected to
go to the training set. And each of these users is characterized by its Age (X-axis)
and its EstimatedSalary (y-axis).

--------------> The RED points are the TRAINING set observations for 
which the DEPENDENT variable, "Purchased" is EQUAL TO ZERO.

The GREEN points are the TRAINING set observations for which the DEPENDENT variable,
"Purchased" is EQUAL TO ONE. <---------------

We represented the way our classifier catches these users by plotting what are
called, THE PREDICTION REGIONS. 

RED region is the one that catches all of the user who do not buy the SUV.
Green region is the region where the classifier catches all of the user that do
buy the SUV.

This is the classifier predictioned COMPARED to the TRUTH.

THE POINT is THE TRUTH, THE REGION is the PREDICTION, hence PREDICTION REGION

The straight line that separates the prediction regions is called the

--------------------------> PREDICTION BOUNDARY <-----------------

AND AND AND the fact that its a straight line is not a coincidence...

its because THATS THE ESSENSE OF LOGISTIC REGRESSION.

If the prediction boundary is a straight line its because our logisticRegression 
classisfer is a LINEAR classifier. 

That means that since we are in 2-Dimensions, because we have two indepdent variables
Then SINCE the logistic regression classifier is a linear classifer,
then the prediction boundary separator can only be a straight line.


If we were in 3-dimensions, it would be straight plane.
Separting two SPACES.

Its always going to be a straight line if your classifier is linear classifier.

However, not so fast, as you can see the green SPOTS in the red REGION and the 
red SPOTS in the green REGION, are the SPOTS where our classifier 
INCORRECTLY PREDICTS.

These incorrect predictions are due specifically to the fact that our classifier is 
a linear classifier and also because our users are not linearly distriubted.

Later on, when you build non-linear classifiers, the prediction boundary 
separator will no longer be a straight line.

File: logistic_regression.R
#Logistic Regression in R

#need to import the dataset and set as current directory 

dataset = read.csv('Social_Network_Ads_for_Classification.csv')
dataset = dataset[, 3:5] #Choosing the columns that we want in our dataset

#Splitting the dataset into the Training Set and Test set
library(caTools) #library used to split into training and test
set.seed(123) #setting the seed so that we get the same numbers for example purposes
split = sample.split(dataset$Purchased, SplitRatio = 0.75) # 75% going to the training set
training_set = subset(dataset, split == TRUE) 
test_set = subset(dataset, split == FALSE)

#now we need to take care feature scaling..for the same reason why we did it
#it in Python, so that in this case, the Estimated Salary and Age are not directly compared
#numerically..it doesn't make sense to actually compare the number 40 and the number 60,000...
# and its easy in R to do this we use the scale function
training_set[,1:2] = scale(training_set[,1:2])
test_set[, 1:2] = scale(test_set[, 1:2])

#End of datapreprocessing; now we are going to build the Logistic Regression Model
classifier = glm(formula = Purchased ~ ., #using glm because our LogReg Classifier is linear
                 family = binomial, # for Logistic regression you have to specify the binomial family
                 data = training_set) 
#Predicting the Test set results
prob_pred = predict(classifier, type = 'response', newdata = test_set[-3])     
#Line 24: We are predicting the test set observations using our classifier which is the logistic
# regression classifier from Line 20. Type = 'response' to get back a vector. newdata=test_set[-3]
# because we want to prediict on our test set...prob_pred is going to return the predicted 
# probabilities that the user will buy the product.
#But we dont want the probabilities, we want the 0 and 1 results.

#so we'll do the following
y_pred = ifelse(prob_pred > 0.5, 1, 0) #changing the probabilities to ifelse.

#Now we are creating the confusion matrix
cm = table( test_set[, 3], y_pred) #first argument is the vector of real values, 
#second is the vector of predictions

#Now we are going to visualize the training set results.
#--------> for information on the graph, see logistic_regression.py <-------------
install.packages('ElemStatLearn')
library(ElemStatLearn)
set = training_set
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 1]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
prob_set = predict(classifier, type = 'response', newdata = grid_set)
y_grid = ifelse(prob_set > 0.5, 1, 0)
plot(set[, -3],
     main = 'Logistic Regression (Training set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))


#Why are we making classifiers?
#The goal here is to classify the right users into the right catgories
#trying to make a classifier that will catch the right users into the right category

set = test_set #shortcut so we dont have to change training_set to test_set everywhere
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01) #doing this so that our points are not 
X2 = seq(min(set[, 2]) - 1, max(set[, 1]) + 1, by = 0.01) #squeezed in the graph. we're building the 
grid_set = expand.grid(X1, X2) #grid here. We're doing this for the age&salary columns
#since grid_set is actually a matrix of the two columns, age and salary,
colnames(grid_set) = c('Age', 'EstimatedSalary') #so here we give a name to the columns of the matrix
#remember c is a function that combines values into a vector or list
prob_set = predict(classifier, type = 'response', newdata = grid_set)

#Now we use the classifier to predict the result of each of the pixel observation points
#the imaginary pixel users
y_grid = ifelse(prob_set > 0.5, 1, 0) #since the predict function returns probabilities, we transform
plot(set[, -3], #it into 0 or 1 results. which returns a vector of predictions of all the points in the
     # the grid
     main = 'Logistic Regression (Test set)', #labeling
     xlab = 'Age', ylab = 'Estimated Salary', #labeling
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato')) #choosing colors
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))

#from confusion matrix: 10 + 7 = 17 incorrect predictions
#and 57 + 26 = 83 correct predictions
#type in cm into the console to view the confusion matrix
